File: templates/python/app/scripts/generator/1.generate_random.py
from moose_lib import task  
from datetime import datetime
from faker import Faker
from app.ingest.models import Foo
import requests

@task
def generate_random():  # The name of your script
    """
    This is a sample task that generates random data for the `Foo` data model and sends it to the ingest API
    """
    fake = Faker()
    for i in range(1000):

        # Prepare request data
        foo = Foo(
            primary_key=fake.uuid4(),
            timestamp=fake.date_time_between(start_date='-1y', end_date='now').timestamp(),
            optional_text=fake.text() if fake.boolean() else None
        )
 
        # POST record to Moose Ingest API
        req = requests.post(
            "http://localhost:4000/ingest/Foo",
            data=foo.model_dump_json().encode('utf-8'),
            headers={'Content-Type': 'application/json'}
        )
        req.raise_for_status()
        
        
    ## Tasks must return a dictionary with the following keys:
    # "task": The name of the task function
    # "data": A dictionary of data to be returned to the workflow
    return {
      "task": "generate_random",
      "data": {
          "completed_at": datetime.now().isoformat()
      }
    }
    
    

File: templates/typescript/app/scripts/generator/1.generateRandom.ts
import { cliLog, TaskDefinition, TaskFunction } from "@514labs/moose-lib";
import { Foo } from "../../ingest/models";
import { faker } from "@faker-js/faker";

// Generate 1000 rows of random Foo data and ingest it into the Foo Data Model
const ingest: TaskFunction = async (input: any) => {
  console.log("Starting data generation...");

  // Generate 1000 rows of random Foo data and ingest it into the Foo Data Model
  for (let i = 0; i < 1000; i++) {
    const foo: Foo = {
      primaryKey: faker.string.uuid(),
      timestamp: faker.date.recent({ days: 365 }).getTime(),
      optionalText: Math.random() < 0.5 ? faker.lorem.text() : undefined,
    };

    try {
      const response = await fetch("http://localhost:4000/ingest/Foo", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(foo),
      });

      if (!response.ok) {
        console.log(
          `Failed to ingest record ${i}: ${response.status} ${response.statusText}`,
        );
      }
    } catch (error) {
      console.log(`Error ingesting record ${i}: ${error}`);
    }

    // Add a small delay to avoid overwhelming the server
    if (i % 100 === 0) {
      console.log(`Ingested ${i} records...`);
      await new Promise((resolve) => setTimeout(resolve, 100));
    }
  }

  console.log("Data generation completed!");
  return {
    task: "ingest",
    data: "success",
  };
};

export default function createTask() {
  return {
    task: ingest,
  } as TaskDefinition;
}
